app/actions/annotations_update.rb

```diff
@@ -16,14 +16,27 @@ def update(resource, annotations, annotation_klass, destroy_nil: true)
             annotation_klass.find(resource_guid: resource.guid, key: key)&.destroy
           end
 
-          annotation = annotation_klass.find_or_create(resource_guid: resource.guid, key: key_name, key_prefix: prefix)
-
-          if value.nil? && destroy_nil # this is delete
-            annotation.try(:destroy)
+          if value.nil? && destroy_nil # Delete Annotation
+            annotation_klass.where(resource_guid: resource.guid, key: key_name).
+              where(Sequel.|({ key_prefix: prefix }, { key_prefix: prefix.to_s })).try(:destroy)
             next
           end
 
-          annotation.update(value: value)
+          begin
+            tries ||= 2
+            annotation_klass.db.transaction(savepoint: true) do
+              annotation = annotation_klass.where(resource_guid: resource.guid, key: key_name.to_s).
+                           where(Sequel.|({ key_prefix: prefix }, { key_prefix: prefix.to_s })).first
+              annotation ||= annotation_klass.create(resource_guid: resource.guid, key: key_name.to_s, key_prefix: prefix.to_s)
+              annotation.update(value: value)
```

Comment:
Guess we can clean this up at some point? Then a comment/reminder would be good

Reply:
It will be part of the linked PR and is part of the same BL think its ok.